PetAdoptions/pethistory/index.js

```diff
@@ -0,0 +1,237 @@
+ */
+exports.handler = async (event, context) => {
+    console.log('Processing adoption history messages:', JSON.stringify(event, null, 2));
+
+    const results = [];
+    const batchItemFailures = [];
+
+    // Process each SQS record
+    for (const record of event.Records) {
+        try {
+            const result = await processAdoptionMessage(record);
+            results.push(result);
+        } catch (error) {
+            console.error('Failed to process record:', record.messageId, error);
+            // Add to batch failures instead of throwing
+            batchItemFailures.push({
+                itemIdentifier: record.messageId
+            });
+        }
+    }
+
+    console.log(`Processed ${results.length} messages successfully, ${batchItemFailures.length} failed`);
+
+    // Return batch failure information for SQS to handle partial failures
+    return {
+        batchItemFailures: batchItemFailures,
+        processedCount: results.length,
+        results: results
+    };
+};
+
+/**
+ * Process a single adoption history message from SQS
+ */
+async function processAdoptionMessage(record) {
+    const messageId = record.messageId;
+    // const receiptHandle = record.receiptHandle;
+
+    console.log(`Processing message ${messageId}`);
+
+    // Parse and validate message
+    let adoptionData;
+    try {
+        adoptionData = JSON.parse(record.body);
+        console.log('Parsed adoption data:', adoptionData);
+    } catch (error) {
+        console.error('Invalid JSON in message:', messageId, error);
+        throw new Error(`Invalid JSON in message ${messageId}: ${error.message}`);
+    }
+    // Validate message schema
+    const { error, value } = adoptionMessageSchema.validate(adoptionData);
+    if (error) {
+        console.error('Schema validation failed for message:', messageId, error.details);
+        throw new Error(`Schema validation failed for message ${messageId}: ${error.message}`);
+    }
+    const validatedData = value;
+    try {
+        console.log(`Processing adoption - Transaction: ${validatedData.transactionId}, Pet: ${validatedData.petId}, User: ${validatedData.userId}`);
+        // Write adoption history to database
+        await writeAdoptionHistoryToDatabase(validatedData);
+
+        console.log(`Successfully processed adoption for pet ${validatedData.petId} by user ${validatedData.userId}`);
+
+        return {
+            messageId: messageId,
+            transactionId: validatedData.transactionId,
+            status: 'success'
+        };
+
+    } catch (error) {
+        console.error(`Failed to process adoption for pet ${validatedData.petId}:`, error);
+        throw error;
+    }
+}
+
+/**
+ * Write adoption history record to PostgreSQL database
+ */
+async function writeAdoptionHistoryToDatabase(adoptionData) {
+    let client;
+
+    try {
+        console.log(`Writing adoption history to database: ${adoptionData.transactionId}`);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when untrusted user input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input using `encodeURIComponent()` or `DOMPurify.sanitize()` before logging.  Learn more - https://cwe.mitre.org/data/definitions/117.html